docs/src/proposals/transactions-v2.md

```diff
@@ -158,15 +157,51 @@ pub struct AddressMap {
   #[serde(with = "short_vec")]
   pub entries: Vec<u8>,
 }
+
+#[derive(Serialize, Deserialize)]
+pub struct CompiledInstructionV2 {
+  /// unchanged
+  pub program_id_index: u8,
+
+  /// unchanged
+  #[serde(with = "short_vec")]
+  pub accounts: Vec<u8>,
+
+  /// Only programs in this list may be invoked by the outer instruction
+  #[serde(with = "short_vec")]
+  pub inner_programs: Vec<u8>,
+
+  /// Dependent accounts do not need to be mapped to the memory of the outer
+  /// instruction's execution context. They are available for passing to inner
+  /// instructions
+  #[serde(with = "short_vec")]
+  pub inner_program_accounts: Vec<u8>,
+
+  /// unchanged
+  #[serde(with = "short_vec")]
+  pub data: Vec<u8>,
+}
 ```
 
+#### Dependent accounts
+
+The new transaction format supports lists of programs and accounts that a
+transaction instruction may use when invoking inner instructions. These programs
+and accounts are separate from transaction instruction account inputs so that
+they do not need to incur the cost of serializing into the outer instruction's
+memory which does need to read or write the account.
+
+For extra user protection, only the programs listed in `inner_programs` may be
+invoked by the outer transaction instruction.
```

Comment:
is this strict requirement? is there any concrete thread in mind? if that's the case, I'll try to address this at #17796 as well.

Reply:
I think I already outlined a general threat here: https://github.com/solana-labs/solana/pull/17901#discussion_r650412745 but more concretely: If a user inadvertently calls a compromised or otherwise malicious program in their transaction, sandboxing can help limit what damage may be done. Transactions can restrict which programs are called and which signers can be used in an instruction to avoid a program being able to invoke a system or token program CPI with a signer account.